File: packages/jetstream_wireformat/src/binary-reader.ts
/**
 * r[impl jetstream.wireformat.ts.reader]
 * r[impl jetstream.wireformat.byte-order]
 */

/**
 * BinaryReader reads from a Uint8Array with a cursor position.
 * All multi-byte reads use little-endian byte order per 9P2000.L conventions.
 */
export class BinaryReader {
  private buffer: Uint8Array;
  private view: DataView;
  private offset: number;

  constructor(buffer: Uint8Array) {
    this.buffer = buffer;
    this.view = new DataView(buffer.buffer, buffer.byteOffset, buffer.byteLength);
    this.offset = 0;
  }

  get remaining(): number {
    return this.buffer.byteLength - this.offset;
  }

  readBytes(count: number): Uint8Array {
    if (this.offset + count > this.buffer.byteLength) {
      throw new Error(`unexpected EOF: wanted ${count} bytes, have ${this.remaining}`);
    }
    const result = this.buffer.slice(this.offset, this.offset + count);
    this.offset += count;
    return result;
  }

  readU8(): number {
    if (this.offset + 1 > this.buffer.byteLength) {
      throw new Error('unexpected EOF: wanted 1 byte');
    }
    const value = this.view.getUint8(this.offset);
    this.offset += 1;
    return value;
  }

  readU16(): number {
    if (this.offset + 2 > this.buffer.byteLength) {
      throw new Error('unexpected EOF: wanted 2 bytes');
    }
    const value = this.view.getUint16(this.offset, true);
    this.offset += 2;
    return value;
  }

  readU32(): number {
    if (this.offset + 4 > this.buffer.byteLength) {
      throw new Error('unexpected EOF: wanted 4 bytes');
    }
    const value = this.view.getUint32(this.offset, true);
    this.offset += 4;
    return value;
  }

  readU64(): bigint {
    if (this.offset + 8 > this.buffer.byteLength) {
      throw new Error('unexpected EOF: wanted 8 bytes');
    }
    const value = this.view.getBigUint64(this.offset, true);
    this.offset += 8;
    return value;
  }

  readI16(): number {
    if (this.offset + 2 > this.buffer.byteLength) {
      throw new Error('unexpected EOF: wanted 2 bytes');
    }
    const value = this.view.getInt16(this.offset, true);
    this.offset += 2;
    return value;
  }

  readI32(): number {
    if (this.offset + 4 > this.buffer.byteLength) {
      throw new Error('unexpected EOF: wanted 4 bytes');
    }
    const value = this.view.getInt32(this.offset, true);
    this.offset += 4;
    return value;
  }

  readI64(): bigint {
    if (this.offset + 8 > this.buffer.byteLength) {
      throw new Error('unexpected EOF: wanted 8 bytes');
    }
    const value = this.view.getBigInt64(this.offset, true);
    this.offset += 8;
    return value;
  }

  readF32(): number {
    if (this.offset + 4 > this.buffer.byteLength) {
      throw new Error('unexpected EOF: wanted 4 bytes');
    }
    const value = this.view.getFloat32(this.offset, true);
    this.offset += 4;
    return value;
  }

  readF64(): number {
    if (this.offset + 8 > this.buffer.byteLength) {
      throw new Error('unexpected EOF: wanted 8 bytes');
    }
    const value = this.view.getFloat64(this.offset, true);
    this.offset += 8;
    return value;
  }
}

File: Sources/JetStreamWireFormat/BinaryReader.swift
// JetStream BinaryReader
// Copyright (c) 2024, Sevki <s@sevki.io>
// SPDX-License-Identifier: BSD-3-Clause

import Foundation

// r[impl jetstream.wireformat.swift.reader]
/// A reader that reads binary data from a `Data` buffer with a cursor position.
public struct BinaryReader {
    private let data: Data
    public private(set) var offset: Int

    /// Creates a new BinaryReader from the given data.
    public init(data: Data) {
        self.data = data
        self.offset = 0
    }

    /// The number of bytes remaining to be read.
    public var remaining: Int {
        return data.count - offset
    }

    /// Reads exactly `count` bytes from the buffer.
    public mutating func readBytes(count: Int) throws -> Data {
        guard offset + count <= data.count else {
            throw WireFormatError.unexpectedEOF
        }
        let result = data[data.startIndex + offset ..< data.startIndex + offset + count]
        offset += count
        return result
    }

    /// Reads a single UInt8.
    public mutating func readU8() throws -> UInt8 {
        guard offset < data.count else {
            throw WireFormatError.unexpectedEOF
        }
        let value = data[data.startIndex + offset]
        offset += 1
        return value
    }

    // r[impl jetstream.wireformat.byte-order]
    /// Reads a UInt16 in little-endian byte order.
    public mutating func readU16() throws -> UInt16 {
        let bytes = try readBytes(count: 2)
        return bytes.withUnsafeBytes { $0.loadUnaligned(as: UInt16.self).littleEndian }
    }

    /// Reads a UInt32 in little-endian byte order.
    public mutating func readU32() throws -> UInt32 {
        let bytes = try readBytes(count: 4)
        return bytes.withUnsafeBytes { $0.loadUnaligned(as: UInt32.self).littleEndian }
    }

    /// Reads a UInt64 in little-endian byte order.
    public mutating func readU64() throws -> UInt64 {
        let bytes = try readBytes(count: 8)
        return bytes.withUnsafeBytes { $0.loadUnaligned(as: UInt64.self).littleEndian }
    }

    /// Reads an Int16 in little-endian byte order.
    public mutating func readI16() throws -> Int16 {
        let bytes = try readBytes(count: 2)
        return bytes.withUnsafeBytes { $0.loadUnaligned(as: Int16.self).littleEndian }
    }

    /// Reads an Int32 in little-endian byte order.
    public mutating func readI32() throws -> Int32 {
        let bytes = try readBytes(count: 4)
        return bytes.withUnsafeBytes { $0.loadUnaligned(as: Int32.self).littleEndian }
    }

    /// Reads an Int64 in little-endian byte order.
    public mutating func readI64() throws -> Int64 {
        let bytes = try readBytes(count: 8)
        return bytes.withUnsafeBytes { $0.loadUnaligned(as: Int64.self).littleEndian }
    }

    /// Reads a Float (IEEE 754 binary32) in little-endian byte order.
    public mutating func readF32() throws -> Float {
        let bits = try readU32()
        return Float(bitPattern: bits)
    }

    /// Reads a Double (IEEE 754 binary64) in little-endian byte order.
    public mutating func readF64() throws -> Double {
        let bits = try readU64()
        return Double(bitPattern: bits)
    }
}
